gpytorch/lazy/kronecker_product_added_diag_lazy_tensor.py

```diff
@@ -0,0 +1,139 @@
+        self.eig_cache = [torch.symeig(lt.evaluate(), eigenvectors = True) for lt in self.lazy_tensors]
+
+    def inv_matmul(self, rhs, jitter=settings.tridiagonal_jitter.value()):
+        Vinv = KroneckerProductLazyTensor(*[DiagLazyTensor(1 / (s[0].abs()+jitter) ) for s in self.eig_cache])
+        Q = KroneckerProductLazyTensor(*[NonLazyTensor(s[1]) for s in self.eig_cache])
+
+        # first compute Q^T y
+        res1 = Q.t().matmul(rhs)
+
+        # now V^{-1} Q^T y
+        res2 = Vinv.matmul(res1)
+        res3 = Q.matmul(res2)
+
+        return res3
+
+    def logdet(self):
+        lt_sizes = [lt.size(-1) for lt in self.lazy_tensors]
+
+        # det(A \kron B) = det(A)^m det(B)^n where m,n are the sizes of A,B
+        scaled_logdets = [m * s[0].sum() for m, s in zip(lt_sizes, self.eig_cache)]
+
+        full_logdet = 0.
+        for logdet in scaled_logdets:
+            full_logdet = logdet + full_logdet
+        
+        return full_logdet
+
+
+class KroneckerProductPlusDiagLazyTensor(AddedDiagLazyTensor):
+    def __init__(self, *lazy_tensors):
+        super(KroneckerProductPlusDiagLazyTensor, self).__init__(*lazy_tensors)
+        if len(lazy_tensors) > 2:
+            raise RuntimeError("An AddedDiagLazyTensor can only have two components")
+        if isinstance(lazy_tensors[0], DiagLazyTensor) and isinstance(lazy_tensors[1], DiagLazyTensor):
+            raise RuntimeError("Trying to lazily add two DiagLazyTensors. " "Create a single DiagLazyTensor instead.")
+        elif isinstance(lazy_tensors[0], DiagLazyTensor):
+            self._diag_tensor = lazy_tensors[0]
+            self._lazy_tensor = lazy_tensors[1]
+        elif isinstance(lazy_tensors[1], DiagLazyTensor):
+            self._diag_tensor = lazy_tensors[1]
+            self._lazy_tensor = lazy_tensors[0]
+        else:
+            raise RuntimeError("One of the LazyTensors input to AddedDiagLazyTensor must be a DiagLazyTensor!")
+
+    def inv_quad_logdet(self, inv_quad_rhs=None, logdet=False, reduce_inv_quad=True):
+        # print("Calling the right inv quad logdet")
+        if inv_quad_rhs is not None:
+            inv_quad_term = self.inv_quad(inv_quad_rhs)
+        else:
+            inv_quad_term = None
+
+        if logdet is not None:
+            logdet_term = self.logdet()
+        else:
+            logdet_term = None
+
+        return inv_quad_term, logdet_term
+
+    def logdet(self):
+        noise = self._diag_tensor[0,0]
+        ### THIS WORKS FOR 2 DIMENSIONS NOW ###
+        sub_eigs = []
+        for lt in self._lazy_tensor.lazy_tensors:
+            sub_eigs.append(lt.evaluate().eig()[0][:, 0].unsqueeze(-1))
+
+        eigs = sub_eigs[0].matmul(sub_eigs[1].t())
+        return torch.log(eigs + noise).sum()
+
+    def _kron_svd(self):
+        return [lt.evaluate().svd() for lt in self._lazy_tensor.lazy_tensors]
+
+    def inv_quad(self, rhs):
+        # TODO: check stability of numerics here
+
+        svd_list = self._kron_svd()
```

Comment:
I can imagine that this could be an issue if this lazy tensor is used with bigger matrices (e.g. if this is used for multi-task models). Can we add a settings flag that peforms this svd-based approach only if the size is sufficiently small?

Reply:
These are pretty outdated now but could be handled by #1102 and #1105 .